File: baekjoon/2667_단지번호붙이기/B_2667_2.js
const dfsRecursion = function(coordinates, count) {
    let y_direction = [-1, 0, 1, 0]
    let x_direction = [0, 1, 0, -1]

    const [y, x] = coordinates
    visited[y][x] = true
    count += 1

    for (let k = 0; k < 4; k++) {
        y_axis = y_direction[k] + y
        x_axis = x_direction[k] + x

        if (0 <= y_axis && y_axis < N && 0 <= x_axis && x_axis < N) {
            if (apartments[y_axis][x_axis] === 1 && visited[y_axis][x_axis] === false) {
                count = dfsRecursion([y_axis, x_axis], count)
            }
        }
    }

    return count
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = Number(input[0])
let apartments = []
for (let i = 1; i <= N; i++) {
    const apartmentInfo = input[i].split('').map(v => Number(v))
    apartments.push(apartmentInfo)
}

let visited = new Array(N).fill(false).map(v => new Array(N).fill(false))

let answer = 0
let counts = []
for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
        if (apartments[i][j] === 1 && visited[i][j] === false) {
            answer += 1
            const count = dfsRecursion([i, j], 0)
            counts.push(count)
        }
    }
}

console.log(answer)
counts.sort((a, b) => a - b)
counts.forEach((item) => console.log(item))

File: baekjoon/2667_단지번호붙이기/B_2667_2.py
import sys


def dfs_recursion(y, x):
    global visited, cnt
    y_axis = [-1, 0, 1, 0]
    x_axis = [0, 1, 0, -1]
    for k in range(4):
        y_index = y_axis[k] + y
        x_index = x_axis[k] + x
        if 0 <= y_index < N and 0 <= x_index < N:
            if visited[y_index][x_index] == 0 and apartments[y_index][x_index] == 1:
                visited[y_index][x_index] = 1
                dfs_recursion(y_index, x_index)
                cnt += 1
    return


N = int(sys.stdin.readline())
apartments = []
for _ in range(N):
    # apartment = list(map(int, input()))
    apartment = list(map(int, sys.stdin.readline().rstrip()))
    apartments.append(apartment)

answer = []
cnt = 1
visited = [[0] * N for _ in range(N)]
for i in range(N):
    for j in range(N):
        if apartments[i][j] == 1 and not visited[i][j]:
            visited[i][j] = 1
            dfs_recursion(i, j)
            answer.append(cnt)
            cnt = 1

answer.sort()
sys.stdout.write(str(len(answer)) + "\n")
for a in answer:
    sys.stdout.write(str(a) + "\n")
